Sources/Foundation/Decimal.swift

```diff
@@ -1687,64 +1687,74 @@ public func NSDecimalString(_ dcm: UnsafePointer<Decimal>, _ locale: Any?) -> St
         return "0"
     }
 
+    var decimalSeparator: String!
+    if let locale = locale as? Locale {
+        decimalSeparator = locale.decimalSeparator
+    } else if let dictionary = locale as? [NSLocale.Key: String] {
+        decimalSeparator = dictionary[NSLocale.Key.decimalSeparator]
+    } else if let dictionary = locale as? [String: String] {
+        decimalSeparator = dictionary[NSLocale.Key.decimalSeparator.rawValue]
+    } else {
+        decimalSeparator = "."
+    }
+    let decimalSeparatorCChar = decimalSeparator.utf8CString // should be accessed with withUnsafeBufferPointer
+    let decimalSeparatorCCharCount = decimalSeparatorCChar.withUnsafeBufferPointer { $0.count } - 1 // without null terminator
+
     let ZERO: CChar = 0x30 // ASCII '0' == 0x30
-    let DECIMALPOINT: CChar = 0x2e // ASCII '.' == 0x2e
     let MINUS: CChar = 0x2d // ASCII '-' == 0x2d
-    let bufferSize = 200 // Max value: 39 + 128 + sign + decimal point
-    var buffer = Array<CChar>(repeating: 0, count: bufferSize)
+    // FIXME: where magic constants 39 and 128 come from?
+    let bufferSize = 39 + 128 + 1 /* sign */ + decimalSeparatorCCharCount + 1 /* null terminator */
 
     var i = bufferSize - 1
-    while copy._exponent > 0 {
-        i -= 1
-        buffer[i] = ZERO
-        copy._exponent -= 1
-    }
-
-    if copy._exponent == 0 {
-        copy._exponent = 1
-    }
+    let buffer = Array<CChar>(unsafeUninitializedCapacity: bufferSize) { buf, initializedCount in
```

Comment:
I wonder if this current algorithm of allocating an array, then adding all of the digits to it backwards and then constructing a string from the result is slightly more than necessary. A better alternative might be to just use a `var result = ""`, append the digits and then return `result.reverse()`.

 If a decent heuristic can be found to estimate the number of output characters then `reserveCapacity()` can be called to just do one allocation. Then in the best case where the result fits into a short string no heap allocation will occur.